pkg/dockerregistry/server/configuration/configuration.go

```diff
@@ -0,0 +1,105 @@
+package configuration
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"os"
+	"reflect"
+	"strings"
+
+	"github.com/docker/distribution/configuration"
+)
+
+type openshiftConfig struct {
+	Openshift Configuration
+}
+
+type Configuration struct {
+	Version configuration.Version `yaml:"version"`
+	Metrics Metrics               `yaml:"metrics"`
+}
+
+type Metrics struct {
+	Enabled bool   `yaml:"enabled"`
+	Secret  string `yaml:"secret"`
+}
+
+func Parse(rd io.Reader) (*configuration.Configuration, *Configuration, error) {
+	in, err := ioutil.ReadAll(rd)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	openshiftEnv, err := popEnv("REGISTRY_OPENSHIFT_")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	dockerConfig, err := configuration.Parse(bytes.NewBuffer(in))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	dockerEnv, err := popEnv("REGISTRY_")
+	if err != nil {
+		return nil, nil, err
+	}
+	if err := pushEnv(openshiftEnv); err != nil {
+		return nil, nil, err
+	}
+
+	config := openshiftConfig{}
+
+	p := configuration.NewParser("registry", []configuration.VersionedParseInfo{
+		{
+			Version: configuration.CurrentVersion,
+			ParseAs: reflect.TypeOf(config),
+			ConversionFunc: func(c interface{}) (interface{}, error) {
+				return c, nil
+			},
+		},
+	})
+
+	if err = p.Parse(in, &config); err != nil {
+		return nil, nil, err
+	}
+	if err := pushEnv(dockerEnv); err != nil {
+		return nil, nil, err
+	}
+
+	return dockerConfig, &config.Openshift, nil
```

Comment:
@miminar I'm not sure I understand the question but I will try to answer. The docker developers  always return the [`configuration.Configuration`](https://github.com/docker/distribution/blob/master/configuration/configuration.go#L615) structure. They can convert old config to this structure, but they can't return something else.